---
title: "Calculating GSVA Scores for TCGA studies: GBM and SKCM"
author: "Sander Tan, The Hyve"
date: "21/04/2017"
output: html_document
---

This documents describes how GSVA scores are calculated from expression data for both GBM and SKCM TCGA studies. In addition to scores, also p-values are calculated with a bootstrapping method that is included in the GSVA package. Staging files are created that are ready to load into cBioPortal.

Note that this requires the version of cBioPortal that includes gene set support (probably 1.6.0) and a database schema of version 2.1.0 or later. 

Also, a manual change has to be made to the meta file of the expression Z-Scores to include `source_stable_id`.

```{r setup, eval=TRUE, message = FALSE, warning = FALSE}
### We want to objects to be saved, for faster kniting.
knitr::opts_chunk$set(cache=TRUE)
knitr::opts_chunk$set(dpi=200)

### Qusage is used to read in a GMT file and parse it as a list
library(qusage)
library(GSVA)
library(biomaRt)
library(reshape2)
library(ggplot2)
library(gridExtra)

### Input files
gbm_expr_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/gbm/tcga/data_RNA_Seq_v2_expression_median.txt"
skcm_expr_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/skcm/tcga/data_RNA_Seq_v2_expression_median.txt"

genesets_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/msigdb.v6.0.entrez.gmt"
result_custom_genesets_file <- "/Users/sander/Data/Project_GSVA/24_TestData2/genesetsHugo2016.entrez.gmt"
result_all_genesets_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/msigdb.v6.0_inclHugo2016.entrez.gmt"

### Output files
skcm_data_scores_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/skcm/tcga/data_gsva_scores.txt"
skcm_data_pvalues_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/skcm/tcga/data_gsva_pvalues.txt"
skcm_meta_scores_file = "/Users/sander/Data/Project_GSVA/26_TestData4/skcm/tcga/meta_gsva_scores.txt"
skcm_meta_pvalues_file = "/Users/sander/Data/Project_GSVA/26_TestData4/skcm/tcga/meta_gsva_pvalues.txt"
skcm_case_list_file = "/Users/sander/Data/Project_GSVA/26_TestData4/skcm/tcga/case_lists/cases_gsva_scores.txt"

### Output files
gbm_data_scores_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/gbm/tcga/data_gsva_scores.txt"
gbm_data_pvalues_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/gbm/tcga/data_gsva_pvalues.txt"
gbm_meta_scores_file = "/Users/sander/Data/Project_GSVA/26_TestData4/gbm/tcga/meta_gsva_scores.txt"
gbm_meta_pvalues_file = "/Users/sander/Data/Project_GSVA/26_TestData4/gbm/tcga/meta_gsva_pvalues.txt"
gbm_case_list_file = "/Users/sander/Data/Project_GSVA/26_TestData4/gbm/tcga/case_lists/cases_gsva_scores.txt"

### Cache files
gsva_skcm_result_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/skcm_tcga_gsva_result.Rdata"
gsva_gbm_result_file <- "/Users/sander/Data/Project_GSVA/26_TestData4/gbm_tcga_gsva_result.Rdata"

### Parameters
n_bootstraps <- 2

### Retrieve versions
gsva_version <- as.character(packageVersion("gsva"))
r_version <- as.character(getRversion())
```

First we'd like to convert the gene sets from mmc2.xlsx (Hugo et al, 2016) to GMT format with Entrez Gene IDs.
```{r}
mapki_genesets_names <- c(
	"MAPKi_INDUCED_EMT",
	"MAPKi_INDUCED_ANGIOGENESIS",
	"EP_BLOOD_VESS_DEVEL_DN_IN_R",
	"PH_BLOOD_VESS_DEVEL_DN_IN_R",
	"EP_RESP_TO_WOUNDING_DN_IN_R",
	"PH_RESP_TO_WOUNDING_DN_IN_R",
	"MS_RESP_TO_WOUNDING_UP_IN_MAPKi_aPDL1_NR",
	"MS_RESP_TO_HYPOXIA_UP_IN_MAPKi_aPDL1_NR",
	"POST_OP_WOUNDHEALING",
	"HOEK_INVASIVE_SIG",
	"HOEK_PROLIFEATIVE_SIG",
	"MAPKR_BLOOD_VESS_DEVEL_UP",
	"DTPP_BLOOD_VESS_DEVEL_UP",
	"DTP_BLOOD_VESS_DEVEL_UP",
	"MAPKR_REG_CELL_PROLIF_UP",
	"DTPP_REG_CELL_PROLIF_UP",
	"DTPP_CELL_ADHESION_UP",
	"PLX2D_CELL_ADHESION_UP",
	"DTPP_RESP_TO_WOUNDING_UP"
)

mapki_genesets_genes <- c(
	"ADAM12,AURKA,BCAT1,BGN,CDH11,CENPF,CKS2,COL10A1,COL11A1,COL3A1,COL5A1,COL5A2,COL6A2,COL6A3,DTL,EPYC,FAP,FCGR1B,FN1,GREM1,IGHM,INHBA,KIF2C,LOXL2,LRRC15,MMP11,NCAPG,NID2,NUAK1,RRM2,SLC16A3,SULF1,TNFAIP6,VCAN",
	"ANPEP,BGN,BMP4,CDH5,COL3A1,CYR61,DLL4,EDN1,EMCN,ID1,KDR,NRP1,PLAU,PPAP2B,PROK2,PRRX2,RHOB,ROBO4,SOX17,SOX18,TGFB2,THBS1,THY1,VEGFA,VEGFC",
	"FGF9,PGF,S100A7,PDGFA,TNFRSF12A,EDN1,ANPEP,PRRX2,SRF,CDH5,TGFB2,SHB,HAND2,HMOX1,ROBO4,RHOB,IL1B,SOX18,SOX17,THBS1,ANGPT2,PPAP2B,CYR61,BMP4,KLF5,FLT1,JUNB,SLIT2,KDR,PROK2,VEGFC,BGN,MEOX2,EREG,ID1,JMJD6,DLL4,VEGFA",
	"EMCN,ACVRL1,LMO2,IL18,COL3A1,FGF10,ANPEP,ENPEP,PRRX2,GJA4,CXCL12,MMP2,CDH5,TGFB2,EDNRA,ACE,S1PR1,PTK2B,TDGF1,ROBO4,PLCD1,SOX18,SOX17,LOX,PPAP2B,COL18A1,BMP4,SELP,EPAS1,EGFL7,TGFBR2,COL15A1,TBX1,KDR,THY1,PROK2,BGN,ID1,PROK1,DLL4,PLXDC1,NOTCH4,ECSCR,COL1A2,ZFPM2,ATPIF1,ENG",
	"F2RL2,S100A8,PDGFB,PDGFA,F2RL1,S100A9,TLR2,CXCR1,CXCR2,IL11,TGFB2,CASP6,FOS,S1PR3,MYD88,LTB4R,HMOX1,SERPINE1,CCL3L3,IL1B,IRAK2,F11,IL18RAP,F10,PLAUR,PROK2,TNFAIP6,THBD,IL20RB,EREG,CARD18,SERPINB2,DSP,RIPK2,KDM6B,NGF,CXCL1,CCL3,CCL2,C9,CXCL3,DRD5,CXCL2,CCL8,BDKRB1,PF4,FPR2,CXCL6,TRIM72,CCL5,SRF,CCL7,TNFRSF1A,PCSK1,MEFV,CCL20,GP1BB,POU2F3,KLKB1,KRT1,CD24,THBS1,PTX3,KLK8,IL6,CEBPB,MAP2K3,S100A12,SOD2,CCL11,ITGA5,HBEGF,ID3,SELE,F2R",
	"F2RL3,ACVRL1,ADORA3,MASP1,TACR1,TGFB3,CXCR1,FGF10,CXCR2,MMRN1,TGFB2,CFHR1,CFP,CASP6,GP5,DYSF,AOAH,CCL3L3,CFH,MS4A2,LOX,LBP,CFD,XCR1,IL18RAP,PTGER3,CCL4L2,SERPING1,CDO1,MECOM,PROK2,SIGLEC1,CCR7,KLRG1,CD36,THBD,CD40LG,SERPINF2,PLA2G7,TFPI,AOC3,CYSLTR1,C6,COL3A1,C1R,PF4,GPR68,C1S,CCL5,IL23A,CCL23,MEFV,GP1BB,CNR2,NFATC4,NOX4,SELP,KL,EFEMP2,TGFBR2,IGF2,C4BPA,CCL16,CCL18,NOTCH3,VWF,ID3,ENG,SELE,IGFBP4,BMP6",
	"ADORA3,PDGFA,TLR2,ADORA1,TGFB1,IL10,CD97,S1PR3,GPX1,NLRC4,MYD88,CCL3L1,SERPINA3,CHST2,STXBP1,CCL4L2,SERPING1,CD40,GAL,PLAUR,C8G,ADM,CTSB,VSIG4,PLA2G2D,TPST1,GGCX,TF,CCL2,RTN4RL1,ADORA2A,CCL8,C1R,ITGB3,FPR2,TIMP3,TNFRSF4,CCL7,TNFRSF1A,SLC11A1,TNFRSF1B,GP1BB,IL10RB,RAC1,SCN9A,ENO3,C2,SCG2,FN1,SPP1,NOX4,PLAT,IL2RA,TNFSF4,STAT3,CCNB1,APOL2,CD55,TFRC,C1RL,SYT17",
	"TF,CCL2,FLT1,ACTN4,SOCS3,PDGFA,ALDOC,EGLN3,NR4A2,TGFB1,DDIT4,CD38,HYOU1,HSP90B1,PLOD1,TFRC,ADM,PLOD2,VEGFA,PSEN2,MT3,ANGPTL4",
	"MMP3,PPBP,CXCL5,PTX3,PTHLH,TDO2,SPINK6,SPP1,MMP10,IL8,MME,GREM1,CTSZ,CXCL6,THBS1,SCG5,TFPI2,PTGS2,CXCL1,IL1A,PCSK1,AREG,IL13RA2,KIAA1199,CCL18,FST,LILRB1,CTNNB1,CLC,CXCL3,CEACAM6,LILRB2,ITCH,S100A12,CCDC102B,GLIS3,MS4A6E,RARRES1,NRG1,PHLDA1,MS4A4A,HAS2,TFEC,CCR1,ANXA3,CR1,IL1RL1,ADAM12,CCNA1,PLA2G7,ENPEP,SPON1,INHBA,STEAP1,STEAP4,TMSB15A,FGF7,PI15,C8orf4,CYBB,MED18,IGSF6,SAA1,RGS13,DEFB4A,SLC16A3,CCL3,AQPEP,CYP1B1,FAM20A,DKK1,IKBIP,SULF1,PXDN,HMOX1,FMO3,SERPINA3,NAA15,MSR1,CCL8,TMEFF1,KLK6,C13orf33,TNFAIP6,MGST1,SRSF6,SRGN,IGF2BP3,PCSK5,LAMC2,OLFML2B,NCEH1,FABP4,IL6,C5AR1,ALDH1A3,PDPN,LYZ,CD163,RAB12,RGS18,HBB,TIMP1,CNN3,FAM83A,CYR61,TNC,DPYSL3,PRR16,BAG2,DSEL,LIPG,PLAC8,CXCL2,FCER1G,SUSD5,NEXN,KLHL6,LMNB1,GPRC5A,TCEAL7,FPR1,APOBEC3A,ITGB6,HS3ST1,GBP6,ITGB5,ADIPOQ,CPXM1,PKP2,NNMT,OLR1,PPP3R1,BUB1,BCL2A1,MAP9,GCLM,S100P,F3,TMPRSS11E,BEND6,FCGR3A,DDX3Y,PI3,MS4A7,FCN1,TLR4,UCHL1,CYTL1,ST8SIA4,MMP9,ALDH1L2,DEPDC1,RNASE2,SPINK7",
	"ADAM12,AMOTL2,AXL,BIRC3,CDH13,CDK14,COL13A1,CRIM1,CRISPLD2,CYR61,DPYD,EFEMP1,EGFR,F2RL1,FGF2,FLNB,FOXD1,FST,FZD2,HEG1,HS3ST3A1,ITGA2,ITGA3,KCNMA1,LOXL2,MYOF,NRP1,NTM,NUAK1,OSMR,PDGFC,PODXL,S100A2,SLC22A4,SLIT2,SYNJ2,TCF4,THBS1,TLE4,TNFRSF11B,TPBG,TPM1,TRAM2,WNT5A,ZEB1",
	"ACP5,ADCY2,APOE,ASAH1,BIRC7,C21orf91,CAPN3,CDH1,CDK2,CDK5R1,CEACAM1,DAPK1,DCT,FAM174B,GALNT3,GNPTAB,GPM6B,GPR143,GPRC5B,GYG2,HPS4,INPP4B,IRF4,IVNS1ABP,KAZ,MBP,MICAL1,MITF,MLANA,MYO1D,NR4A3,OCA2,PHACTR1,PIR,PLXNC1,PMEL,RAB27A,RAB38,RGS20,RHOQ,RRAGD,SEMA6A,SIRPA,SLC45A2,ST3GAL6,STX7,TNFRSF14,TRPM1,TYR,TYRP1,WDR91,ZFYVE16",
	"CAV1,NRP1,EPAS1,COL3A1,EDN1,COL5A1,CITED2,CDH13,VEGFC,S1PR1,JUN,CCBE1,PLCD3,FOXC2,COL1A1,FGF2,PLAU,CYR61",
	"CAV1,NRP1,LMO2,EDN1,COL3A1,MMP2,CXCL12,CITED2,AGT,CCBE1,PLCD3,RHOB,SEMA3C,THBS1,FGF2,CEACAM1,SCG2,CYR61,BMP4,COL18A1,EPAS1,MMP19,MYH9,ARHGAP24,COL5A1,THY1,CDH13,VEGFC,BGN,EPGN,JUN,VEGFA,NTRK2,COL1A2,COL1A1,PLAU",
	"CAV1,NRP1,LMO2,EDN1,COL3A1,TNFSF12,MMP2,CITED2,ANGPTL6,CXCR4,PLCD3,RHOB,QKI,SEMA3C,THBS1,CYR61,KLF5,COL18A1,EPAS1,MMP19,MYH9,ARHGAP24,COL5A1,ANXA2,THY1,SMO,CDH13,BGN,JUN,NTRK2,COL1A2,COL1A1,PLAU",
	"CAV2,RBP4,CAV1,FOSL2,CCL2,NRP1,IGFBP7,CLU,EDN1,NFKBIA,IL15,IL34,SOX9,S1PR3,AGTR1,BDNF,S1PR1,SPEG,HLX,SERPINE1,NKX3-1,PDGFC,CD24,NRG1,RUNX2,FGF2,EGFR,PTGER2,TP53I11,IL6,IRS1,PDCD1LG2,MXD4,VEGFC,CDH13,TNFRSF9,ADRB2,JUN,F3,BNC1,IL12A,PDGFRB,TGFB1I1,PLAU,NGF",
	"RARRES3,FOSL2,FGF7,NRP1,PDGFB,FGF17,IGFBP7,EDN1,GJA1,FOXO4,CXADR,VIPR1,GLI3,IL31RA,AZGP1,AGTR1,WISP2,BDNF,GPC3,HLX,SERPINE1,PDGFC,NRG1,FGF2,EBI3,EGFR,PRKCA,PTPRK,PTGER2,CD40,IRS1,PDCD1LG2,MXD4,MYCN,TNS3,VEGFC,ADRB2,ADAMTS8,CCND2,CHRM1,F3,JUN,BTG4,GRN,VEGFA,IL12A,PDGFRB,NGFR,TGFB1I1,PMP22,NGF,CAV2,CAV1,CCL2,IFITM1,CLU,PTH1R,NFKBIA,KIT,BDKRB2,IL34,TIMP2,SOX9,ADA,VDR,IL12RB1,SPEG,AGT,ADRA2A,NKX3-1, CD24,THBS1,PPAP2A,RUNX2,SCG2,BMP4,COL18A1,IL6,TP53I11,TNFSF4,KAT2B,HCLS1,KLF11,TAX1BP3,CDH13,ATF3,NUPR1,EPGN,ETS1,DLX5,BNC1,FABP4,NR5A2,PLAU,KCTD11,F2R",
	"CADM3,NRP1,THRA,CADM1,TLN2,IGFBP7,NPNT,FERMT2,BCAM,L1CAM,EDIL3,CXADR,CXCL12,VCL,NRCAM,AZGP1,WISP2,TGFBI,RHOB,LOXL2,NEGR1,BOC,CEACAM1,CDH24,CYR61,SPON1,EGFR,F11R,PTPRK,PCDHB7,NRXN2,CNTN6,SDK1,CPXM2,MYH9,THY1,JUP,CD36,CLDN1,LAMC2,TGFB1I1,PARVA,ACHE,CCL2,COL3A1,ITGA11,COL28A1,SPOCK1,IL32,CDH3,PCDHB11,SOX9,APLP1,ALCAM,COL17A1,LAMB2,SORBS1,FAT4,AGT,TTYH1,COL6A3,MSLN,CD24,THBS1,SELPLG,THBS3,APBA1,NPHP1,FN1,COL18A1,FLRT1,BGLAP,BMP1,COL13A1,ITGA1,NFASC,HSPG2,CELSR2,NID2,PCDH17,COL5A3,COL16A1,COL5A1,COL4A6,CDH13,ERBB2IP,ITGA5,PKP3,ADAM22,NTM,FEZ1",
	"THRA,TLN2,FERMT2,L1CAM,BCAM,EDIL3,VCL,AZGP1,WISP2,RHOB,LOXL2,COL11A1,CYR61,PCDHB5,CNTN6,MYH9,JUP,NCAM2,CD36,CD99L2,TGFB1I1,PARVA,COL3A1,PCDHB15,NINJ1,PCDHB11,PKD1L1,CLDN14,ALCAM,SORBS1,ROPN1B,TTYH1,PVRL2,MSLN,ACAN,CD24,THBS1,GPNMB,APBA1,THBS3,COL18A1,MAG,FLRT1,ADAM23,ITGA1,HSPG2,CELSR3,NID2,PCDH17,COL16A1,COL5A3,PCDH18,COL14A1,FREM2,CDH19,CYFIP2,ANTXR1,ABL2",
	"F2RL2,NRP1,FGF7,PDGFB,F2RL1,TLR3,DYSF,SERPINE1,CFH,NRG1,FGF2,IRAK2,F11R,LY96,CD40,SDC1,CD36,F3,TFPI,NGFR,NGF,ACHE,CCL2,RTN4RL1,C3,CXCL3,COL3A1,CXCL2,CLU,C1R,BDKRB2,C1S,CDH3,MDK,TPM1,CCL26,IGSF10,LAMB2,NFATC4,CD24,THBS1,FN1,SCG2,IL6,TNFSF4,EFEMP2,MSTN,COL5A1,APOL3,PLSCR4,NUPR1,ITGA5,CD59,AOX1,PLA2G4C,HDAC9,PLAU,IGFBP4,F2R"
)
mapki_genesets_genes_vector <- strsplit(mapki_genesets_genes, ",")

mart <- useMart("ensembl")
mart <- useDataset("hsapiens_gene_ensembl", mart)

### Retrieve Entrez Gene Ids. Some genes are lost because no Entrez Gene Id is found with HUGO symbol
mapki_genesets_genes_list <- getBMlist(attributes= "entrezgene",
      filters = c("hgnc_symbol"),
      values = c(mapki_genesets_genes_vector), mart)

mapki_genesets_genes_list <- mapki_genesets_genes_list$entrezgene
names(mapki_genesets_genes_list) <- mapki_genesets_names

### Read in gene sets
geneset_inp <- read.gmt(genesets_file)

genesets_all <- c(geneset_inp, mapki_genesets_genes_list)

### Add new gene sets to a gmt file
write_gmt <- function(x, output_name) {
  if (file.exists(output_name)) {
    file.remove(output_name)
  }
  z <- deparse(substitute(x))
  nams=names(x) 
  for (i in seq_along(x)) {
    geneset_vector <- unlist(x[i])
    cat(nams[i],
        "\thttp://www.thehyve.nl\t",
        paste(unlist(x[i]), collapse = "\t"),
        "\n",
        file = output_name,
        append=TRUE,
        sep = "")
  } 
}
### Write custom gene sets
write_gmt(mapki_genesets_genes_list, result_custom_genesets_file)

### Copy original MSigDB gmt
file.copy(genesets_file, result_all_genesets_file, overwrite = TRUE)

### Append custom gene sets to original MSigDB gmt
file.append(result_all_genesets_file, result_custom_genesets_file)
```

# SKCM

```{r}
### Read in expression
skcm_inp <- read.table(skcm_expr_file, header = T, sep = "\t", quote = "", fill = T, check.names = F)
skcm_inp$Entrez_Gene_Id <- as.factor(skcm_inp$Entrez_Gene_Id)

### Remove NA entrez IDs
skcm_expr_not_na <- skcm_inp[complete.cases(skcm_inp),]

### Remove dup
skcm_expr_not_dub <- skcm_expr_not_na[!duplicated(skcm_expr_not_na$Entrez_Gene_Id),]

### Create only expr data table
skcm_expr <- as.matrix(skcm_expr_not_dub[,3:ncol(skcm_expr_not_dub)])
rownames(skcm_expr) <- skcm_expr_not_dub$Entrez_Gene_Id

### Log2 normalize
skcm_expr_norm <- log2(skcm_expr + 1)

### Plot distrubution of reads
skcm_expr_norm_m <- melt(skcm_expr_norm)
colnames(skcm_expr_norm_m) <- c("Gene", "Sample", "Log2_Expression")
p1 <- ggplot(skcm_expr_norm_m, aes(x = Log2_Expression)) + geom_histogram(binwidth = 1) + ggtitle("All genes")

### Remove lowly expressed genes. See:
### https://support.bioconductor.org/p/95407/#95547
mexp <- rowMeans(skcm_expr_norm)
skcm_expr_norm_high <- skcm_expr_norm[mexp > 5, ]

### Plot distrubution of reads
skcm_expr_norm_high_m <- melt(skcm_expr_norm_high)
colnames(skcm_expr_norm_high_m) <- c("Gene", "Sample", "Log2_Expression")
p2 <- ggplot(skcm_expr_norm_high_m, aes(x = Log2_Expression)) + geom_histogram(binwidth = 1) + ggtitle("Lowly expressed genes removed")

### Print the plots
grid.arrange(p1, p2, ncol = 1, top = "TCGA SKCM RNA-Seq V2")
```

### Run GSVA

```{r}
### Run GSVA
#gsva_result <- gsva(skcm_expr_norm, genesets_all, method = "gsva", no.bootstraps = n_bootstraps)
#gsva_result <- gsva(skcm_expr_norm, genesets_all, method = "gsva")
#save(gsva_result, file = gsva_skcm_result_file)
load(gsva_skcm_result_file)
```

### Data files

```{r}
### Extract results and add column for row names
gsva_scores <- data.frame("geneset_id" = rownames(gsva_result$es.obs), gsva_result$es.obs, check.names = F)

### Bootstrapping can be enabled but takes significant amount of time. We create a placeholder table in which all scores have a p-value of 0.01
#gsva_pvalues <- data.frame("geneset_id" = rownames(gsva_result$bootstrap$p.vals.sign), gsva_result$bootstrap$p.vals.sign, check.names = F)
fake_pvalues <- (gsva_result$es.obs * 0) + 0.01
gsva_pvalues <- data.frame("geneset_id" = rownames(gsva_result$es.obs), fake_pvalues, check.names = F)

### Write data files
write.table(gsva_scores, skcm_data_scores_file, quote = F, sep = "\t", col.names = T, row.names = F)
write.table(gsva_pvalues, skcm_data_pvalues_file, quote = F, sep = "\t", col.names = T, row.names = F)
```

### Meta files

```{r}
### Write meta files
meta_scores <- paste0("cancer_study_identifier: skcm_tcga
genetic_alteration_type: GENESET_SCORE
datatype: GSVA-SCORE
stable_id: gsva_scores
source_stable_id: rna_seq_v2_mrna
profile_name: GSVA scores
profile_description: GSVA scores for MSigDB v6.0 genesets based on RNA-Seq V2 calculated with GSVA version ", gsva_version,", R version ", r_version, "
data_filename: data_gsva_scores.txt
geneset_def_version: 6.0")
write(meta_scores, skcm_meta_scores_file)

# meta_pvalues <- paste0("cancer_study_identifier: skcm_tcga
# genetic_alteration_type: GENESET_SCORE
# datatype: P-VALUE
# stable_id: gsva_pvalues
# source_stable_id: gsva_scores
# profile_name: GSVA p-values
# profile_description: P-values calculated for GSVA scores with bootstrapping method (", n_bootstraps, " resamples) using GSVA version ", gsva_version,", R version ", r_version, "
# data_filename: data_gsva_pvalues.txt
# geneset_def_version: 6.0")
# write(meta_pvalues, skcm_meta_pvalues_file)

meta_pvalues <- paste0("cancer_study_identifier: skcm_tcga
genetic_alteration_type: GENESET_SCORE
datatype: P-VALUE
stable_id: gsva_pvalues
source_stable_id: gsva_scores
profile_name: GSVA p-values
profile_description: Mock p-values, these can be calculated with GSVA bootstrapping method.
data_filename: data_gsva_pvalues.txt
geneset_def_version: 6.0")
write(meta_pvalues, skcm_meta_pvalues_file)
```

### Case list

```{r}
### Add case list
case_list <- paste0(c("cancer_study_identifier: skcm_tcga
stable_id: skcm_tcga_gsva_scores
case_list_name: Tumor Samples with GSVA data
case_list_description: All samples with GSVA data
case_list_category: all_cases_with_gsva_data
case_list_ids:    ", paste0(colnames(gsva_result$es.obs), collapse = "\t")), collapse = "")
write(case_list, skcm_case_list_file)
```
Furthermore, we have to make a manual change to the meta file of the expression Z-Scores to add:
source_stable_id: rna_seq_v2_mrna

# GBM

Do same analysis for Glioblastoma study

```{r}
### Read in expression
gbm_inp <- read.table(gbm_expr_file, header = T, sep = "\t", quote = "", fill = T, check.names = F)
gbm_inp$Entrez_Gene_Id <- as.factor(gbm_inp$Entrez_Gene_Id)

### Remove NA entrez IDs
gbm_expr_not_na <- gbm_inp[complete.cases(gbm_inp),]

### Remove dup
gbm_expr_not_dub <- gbm_expr_not_na[!duplicated(gbm_expr_not_na$Entrez_Gene_Id),]

### Create only expr data table
gbm_expr <- as.matrix(gbm_expr_not_dub[,3:ncol(gbm_expr_not_dub)])
rownames(gbm_expr) <- gbm_expr_not_dub$Entrez_Gene_Id

### Log2 normalize
gbm_expr_norm <- log2(gbm_expr + 1)

### Plot distrubution of reads
gbm_expr_norm_m <- melt(gbm_expr_norm)
colnames(gbm_expr_norm_m) <- c("Gene", "Sample", "Log2_Expression")
p1 <- ggplot(gbm_expr_norm_m, aes(x = Log2_Expression)) + geom_histogram(binwidth = 1) + ggtitle("All genes")

### Remove lowly expressed genes. See:
### https://support.bioconductor.org/p/95407/#95547
mexp <- rowMeans(gbm_expr_norm)
gbm_expr_norm_high <- gbm_expr_norm[mexp > 5, ]

### Plot distrubution of reads
gbm_expr_norm_high_m <- melt(gbm_expr_norm_high)
colnames(gbm_expr_norm_high_m) <- c("Gene", "Sample", "Log2_Expression")
p2 <- ggplot(gbm_expr_norm_high_m, aes(x = Log2_Expression)) + geom_histogram(binwidth = 1) + ggtitle("Lowly expressed genes removed")

### Print the plots
grid.arrange(p1, p2, ncol = 1, top = "TCGA GBM RNA-Seq V2")
```

### Run GSVA

```{r}
### Run GSVA
#gsva_result <- gsva(skcm_expr_norm, genesets_all, method = "gsva", no.bootstraps = n_bootstraps)
#gsva_result <- gsva(gbm_expr_norm, genesets_all, method = "gsva")

#save(gsva_result, file = gsva_gbm_result_file)
load(gsva_gbm_result_file)
```

### Data files

```{r}
### Extract results and add column for row names
gsva_scores <- data.frame("geneset_id" = rownames(gsva_result$es.obs), gsva_result$es.obs, check.names = F)

### Bootstrapping can be enabled but takes significant amount of time. We create a placeholder table in which all scores have a p-value of 0.01
#gsva_pvalues <- data.frame("geneset_id" = rownames(gsva_result$bootstrap$p.vals.sign), gsva_result$bootstrap$p.vals.sign, check.names = F)
fake_pvalues <- (gsva_result$es.obs * 0) + 0.01
gsva_pvalues <- data.frame("geneset_id" = rownames(gsva_result$es.obs), fake_pvalues, check.names = F)

### Write data files
write.table(gsva_scores, gbm_data_scores_file, quote = F, sep = "\t", col.names = T, row.names = F)
write.table(gsva_pvalues, gbm_data_pvalues_file, quote = F, sep = "\t", col.names = T, row.names = F)
```

### Meta files

```{r}
### Write meta files
gsva_version <- as.character(packageVersion("gsva"))
r_version <- as.character(getRversion())
meta_scores <- paste0("cancer_study_identifier: gbm_tcga
genetic_alteration_type: GENESET_SCORE
datatype: GSVA-SCORE
stable_id: gsva_scores
source_stable_id: rna_seq_v2_mrna
profile_name: GSVA scores
profile_description: GSVA scores for MSigDB v6.0 genesets based on RNA-Seq V2 calculated with GSVA version ", gsva_version,", R version ", r_version, "
data_filename: data_gsva_scores.txt
geneset_def_version: 6.0")
write(meta_scores, gbm_meta_scores_file)

# meta_pvalues <- paste0("cancer_study_identifier: gbm_tcga
# genetic_alteration_type: GENESET_SCORE
# datatype: P-VALUE
# stable_id: gsva_pvalues
# source_stable_id: gsva_scores
# profile_name: GSVA p-values
# profile_description: P-values calculated for GSVA scores with bootstrapping method (", n_bootstraps, " resamples) using GSVA version ", gsva_version,", R version ", r_version, "
# data_filename: data_gsva_pvalues.txt
# geneset_def_version: 6.0")
# write(meta_pvalues, gbm_meta_pvalues_file)

meta_pvalues <- paste0("cancer_study_identifier: gbm_tcga
genetic_alteration_type: GENESET_SCORE
datatype: P-VALUE
stable_id: gsva_pvalues
source_stable_id: gsva_scores
profile_name: GSVA p-values
profile_description: Mock p-values, these can be calculated with GSVA boostrapping method.
data_filename: data_gsva_pvalues.txt
geneset_def_version: 6.0")
write(meta_pvalues, gbm_meta_pvalues_file)
```

### Case list

```{r}
### Add case list
case_list <- paste0(c("cancer_study_identifier: gbm_tcga
stable_id: gbm_tcga_gsva_scores
case_list_name: Tumor Samples with GSVA data
case_list_description: All samples with GSVA data
case_list_category: all_cases_with_gsva_data
case_list_ids:    ", paste0(colnames(gsva_result$es.obs), collapse = "\t")), collapse = "")
write(case_list, gbm_case_list_file)
```

Furthermore, we have to make a manual change to the meta file of the expression Z-Scores to add:
source_stable_id: rna_seq_v2_mrna

```{r}
print(sessionInfo())
```